### ISSUE TITLE
Specify job name prefix

### ISSUE THREAD

**Issue by gaow** [rfeng2023/mmcloud#27]
This must be some feature that `float` already has and perhaps we just need an interface on mm_jobman.sh : currently our job ID are  random strings, eg: `p2276jrcwy13ytv6e5qnj` it would be nice if we can specify part of it for example `--job-prefix some_analysis` then the job ID show up as `some_analysis_p2276jrcwy13ytv6e5qnj`. @Ashley-Tung is that some available features we can adopt?

**Comment by gaow** [comment]
This will makes filtering and checking the jobs easier.

**Comment by Ashley-Tung** [comment]
This can be done with `-n <string>` in the float command. I can implement this fairly easily in the script. Is there a default name you would like to have?

**Comment by gaow** [comment]
Thanks @Ashley-Tung that is great! we dont need a default. If it is not specified, we dont set `-n` in float so the default will become the float default.

**Comment by gaow** [comment]
Actually, more generally speaking, do you think we can allow for any other `float` features if they are not part of the `mm_jobman.sh` parameters? In other words, things like `-c` `-m` we overwrite. But others like `-n` that we did not define, we can pass that to `float` by default. Just a suggestion as a more general solution. Not a hard request.

**Comment by Ashley-Tung** [comment]
This can be done, but that is up to you to decide what features we can allow by default. Here is a comprehensive list from the `float submit --help` command
```
      --allowList stringArray       the list of instance types able to be used
      --bandwidth int               bandwidth of public network interface required for the job, unit Mbps. (Only valid for AliCloud)
  -A, --cmdArgs string              cmd args to run workload. The arguments can be used to run container directly or provided to job file directly.
  -c, --cpu string                  required CPU cores to use in this form 'minCPU:maxCPU', ':maxCPU' is omittable.
      --cpuVendor string            vendor of CPU. This is used to limit special instance types from cloud provider.
                                        Valid options: intel, amd. (default "", means cloud provider can provide instances with any CPU vendor). (AWS only)
      --customTag stringArray       define custom tags for resources like this: tag1:value1 tag2
  -D, --dataVolume stringArray      data volumes used to run the job. Format is like this:
                                        vol-id:/mnt/path, use an existing volume
                                        [size=100,throughput=100]:/data, create a new volume with spec
                                        nfs://172.31.2.120/export_dir:/mounted_dir, use NFS path
                                        [accesskey=xxx,secret=yyy,endpoint=zzz,mode=rw]oss://bucketname/path:/mountpoint, use OSS bucket, where access key and secret are optional if RAM role is used
                                        [accesskey=xxx,secret=yyy,endpoint=zzz,mode=rw]s3://bucketname/path:/mountpoint, use S3 bucket, where access key and secret are optional if IAM role is used
  -d, --def string                  definition file which defines job execution fields in yaml or json
      --denyList stringArray        the list of instance types need be skipped
      --dumpMode string             dump mode for job, valid options: full, incremental (default "full")
  -e, --env stringArray             provide environment variables for the job like this: env_key=env_value
  -E, --errPolicy string            policy for job failure.
                                        Supported policy (default: reclaimAll):
                                          retainVolumes, (when job failed, keep all volumes. Customer can cancel or resume the job with retained volumes later.)
                                          restart, (when migration failed, restart the job automatically.)
                                          autoSuspend, (when OOM migration failed with no instance type, suspend job for manual resume)
                                          reclaimAll. (when migration failed, mark job failed and reclaim all resources.)
      --extraContainerOpts string   specify extra options for container. The options will be passed to container directly.
  -f, --force                       automatically answer yes for confirmation prompt
      --gateway string              the gateway to connect, specify 'auto' to have the system choose an option for you
  -h, --help                        help for submit
  -i, --image string                specify a image name or URI to run, if URI is specified, it will pick the first image that URI matched
      --imageVolSize int            define image volume size (in GB)
  -t, --instType string             instance type to execute job, like 'c4.xlarge' in AWS or 'ecs.g7.6xlarge' in AliCloud
      --interruptInterval string    interval to interrupt workload in format: 1h2m3s. (default: 0, which means disabled).
      --interruptMaxCount string    max count to interrupt workload, 0 means no limit. (default: 0, which means disabled).
  -j, --job string                  job file to execute. Support local file, S3 file, OSS file and https|http file.
  -m, --mem string                  required memory to use (in GB) in this form 'minMem:maxMem', ':maxMem' is omittable.
      --metricsInterval string      specify interval to collect container metrics in format: 1h2m3s. (default: 10s).
  -M, --migratePolicy string        automatic migration policy. (default enabled with cpu.disable=true,mem.disable=true,evadeOOM=true)
                                        Format is like this (all numbers without unit means percentage):
                                        [disable=true], disable auto migration
                                        [cpu.upperBoundRatio=90,cpu.upperBoundDuration=10s], define cpu related policy only
                                        [cpu.lowerBoundRatio=30,mem.upperBoundRatio=90], define cpu lower bound ratio to 30% and memory upper bound ratio to 90%
                                        [cpu.step=50], define cpu migrate step to 50%
                                        [stepAuto=true], request server to calculate migrate step intelligently
                                        The specified fields will overwrite default policy fields configured globally.
                                        All supported fields are:
                                        - disable, enable, evadeOOM, stepAuto,
                                        - cpu.disable, cpu.step, cpu.limit, cpu.lowerLimit,
                                        - cpu.upperBoundRatio, cpu.upperBoundDuration, cpu.lowerBoundRatio, cpu.lowerBoundDuration,
                                        - mem.disable, mem.step, mem.limit, mem.lowerLimit,
                                        - mem.upperBoundRatio, mem.upperBoundDuration, mem.lowerBoundRatio, mem.lowerBoundDuration
  -n, --name string                 specify the job name
      --noPublicIP                  no public IP required for the job. Please make sure cloud service is accessible on instance without public IP.
  -o, --output string               specify folder to save job output. Please make sure this path is accessible by worker node.
                                        Output file will be saved to this folder with name: stdout.autosave.$jobid stderr.autosave.$jobid.
                                        Caller should take responsibility to delete those files when they are not used.
                                        Supported pattern is:
                                                nfs://server_id/remote_folder,
                                                file:///local_folder or /local_folder (the path must has been mounted to the worker node)
  -P, --publish stringArray         provide container publish for the job like this:
                                        host_port:container_port
                                        Please make sure:
                                        1. the port in container is open
                                        2. the port has been opened in security-group
      --rootVolSize int32           worker node volume size of the disk for root file-system (default 40g)
      --securityGroup stringArray   security groups (or network tags in GCP) to add to the instance for the job
      --shmSize string              size of /dev/shm, it is a number with a unit. (default 64m)
                                        a unit can be b (bytes), k (kibibytes), m (mebibytes), or g (gibibytes)
  -I, --snapshotInterval string     set periodic snapshot interval for job in format: 1h2m3s. (default: 0, which means disabled).
  -s, --subnet stringArray          specify subnet(aws)|vSwitch(ali) to use. (by default, system will choose one for you)
      --tag string                  image tag to run the job (default latest or the only one available tag)
      --targetPort stringArray      the port to connect to gateway
  -l, --timeLimit string            set time limit for job in format: 1h2m3s (default: 0, which means no limit). Job running longer than the limit will be cancelled automatically.
      --trackDirty                  keep track of dirty page size of the job
  -V, --vmPolicy string             VM creation policy. Format is like this:
                                        [spotOnly=true,retryLimit=10,retryInterval=30s], use spot VM only
                                        [spotFirst=true,retryLimit=3], try spot VM firstly, if failed, use on-demand VM
                                        [onDemand=true], use on-demand VM
                                        [spotOnly=true,priceLimit=0.1] declare price limit for spot VMs.
                                        All supported fields are:
                                        - spotOnly, spotFirst, onDemand, retryLimit, retryInterval, priceLimit
                                        (default: spotFirst=true,retryLimit=3,retryInterval=10m)
      --withRoot                    run workload with root directly. This is useful when image has special users in its container.
  -z, --zone string                 availability zone in which to run job

```

**Comment by gaow** [comment]
@Ashley-Tung how many of these parameters we already included in our interface as is? I guess maybe `--image` is directly passed into the `float` command. And, how many we overwrite? Perhaps just `-c`? I think a general idea would be we drop anything we directly pass into it, only keep the ones that we overwrite. Then for all other unspecified command we just leave for `float` to handle? In Python, it's the  `def foo(, **kwargs)` (if my memory still serves for my rusty Python syntax). In R it's `foo(, ...)`

**Comment by gaow** [comment]
I think the point is to avoid overriding as much as possible. I suggest we keep only two types of parameters on `mm_jobman.sh` :  

1. Those that are specific to `mm_jobman.sh`, which cannot be found in `float`, for example, `--job-size`
2. Those that are the same and will override `float`, but are required parameter for `float` anyways without which there will be an error. So we can also allow that in `mm_jobman.sh` to parse and check, before we generate a bunch of wrong float commands. I guess such as `--image`.

For the optional parameters to `float` that we currently directly override with `mm_jobman.sh`, we can clean them up to lighten the mm_jobman interface and direct user attention to `float`.

@Ashley-Tung i think you and i are on the same page about it anyways. Just want to make sure

**Comment by Ashley-Tung** [comment]
That sounds good to me. We'd have to go through the list of parameters to make sure we are confident in our choices. As of now, and with my latest pull request #29 , the parameters are
```
    echo "  -c <min>:<optional max>                   Specify the min and max of CPUs to use (default and recommended for AWS Spot Instances: 2)."
    echo "  -m <min>:<optional max>                   Set the min and max of memory in GB (default: 16)."
    echo "  -n <name>                                 Name of job. Leave empty for randomly-generated string (optional)."
    echo "  --cwd <value>                             Define the working directory for the job (default: ~)."
    echo "  --download <remote>:<local>               Download files/folders from S3. Format: <S3 path>:<local path> (optional)."
    echo "  --upload <local>:<remote>                 Upload folders to S3. Format: <local path>:<S3 path> (optional)."
    echo "  --download-include '<value>'              Use the include flag to include certain files for download (space-separated) (optional)."
    echo "  --dryrun                                  Execute a dry run, printing commands without running them."
    echo "  --entrypoint '<command>'                  Set the initial command to run in the job script (required)."
    echo "  --env <key>=<val>                         Set environmental variables for the job in the format KEY=VALUE (optional)."
    echo "  --image <value>                           Specify the Docker image to use for the job (required)."
    echo "  --imageVolSize <value>                    Define the size of the image volume in GB (depends on the size of input image)."
    echo "  --job-size <value>                        Set the number of commands per job for creating virtual machines (default: 2)."
    echo "  --mount <bucket>:<local>                  Mount an S3 bucket to a local directory. Format: <bucket>:<local path> (optional)."
    echo "  --mountOpt <value>                        Specify mount options for the bucket (required)."
    echo "  --ebs-mount <folder>=<size>               Mount an EBS volume to a local directory. Format: <local path>=<size>. Size in GB (optional)."
    echo "  --no-fail-fast                            Continue executing subsequent commands even if one fails."
    echo "  --opcenter <value>                        Provide the Opcenter address for the job (required)."
    echo "  --parallel-commands <value>               Set the number of commands to run in parallel (default: number of CPUs)."
```

**Comment by gaow** [comment]
looks like we can get rid of --env, --image, --imageVolSize? Others will have to stay, based on the two rules i proposed earlier.

**Comment by Ashley-Tung** [comment]
Just want to make sure, if we get rid of them, we can just make sure our float command just always has those values?
For example, we needed `env` for `ENV_NAME=pecotmr`, `--image ghcr.io/cumc/pecotmr_docker:latest`, and `--imageVolSize 10` in order for the job to run as expected?

**Comment by gaow** [comment]
Those can be passed into float still though mm_jobman.sh right? It's just for whatever options that mm_jobman doesn't process, they go directly into float. Are you familiar with python**kwarg? It's the same ide

**Comment by Ashley-Tung** [comment]
I am not too familiar, but is what you are saying is once we simplify `mm_jobman`, the user still has the option to use more parameters, but instead of parsing them in the script, we just add them to the end of our submitted float command?

**Comment by gaow** [comment]
>the user still has the option to use more parameters, but instead of parsing them in the script, we just add them to the end of our submitted float command?

Exactly! Whatever mm_jobman.sh parser does not recognize, we trust that users know what they are doing and simply attach them as is to the end of `float` command
